### PR TITLE
Fix for option 1 of JournalDB

### PR DESCRIPTION
**Pull request by gavofyork**

### CODE REVIEW

File: util/src/journaldb.rs [hunk 1]
```diff
@@ -42,7 +71,8 @@ impl Clone for JournalDB {
 		JournalDB {
 			overlay: MemoryDB::new(),
 			backing: self.backing.clone(),
-			counters: self.counters.clone(),
+			refs: self.refs.clone(),
+			last_era: 0,
```
Comment (tomusdrw) [inline util/src/journaldb.rs]: Why 0 here? Isn't it related to what's in DB?
(it's not read anyway)

Comment (gavofyork) [inline util/src/journaldb.rs]: old debug code. removing.

**Comment by arkpar** [comment]
DB version bump needed

File: util/src/journaldb.rs [hunk 2]
```diff
@@ -24,6 +24,34 @@ use kvdb::{Database, DBTransaction, DatabaseConfig};
 #[cfg(test)]
 use std::env;
 
+#[derive(Clone, PartialEq, Eq)]
+struct RefInfo {
```
Comment (arkpar) [inline util/src/journaldb.rs]: Idea for future optimization: compact this struct. Too bad rust does not support bit fields. In my original implementation I've used `u16` for counter and `#[repr(packed)]` for the structure to reduce memory usage. It is also possible to use a single bit of `queue_refs` for the flag

File: util/src/journaldb.rs [hunk 3]
```diff
@@ -170,7 +200,10 @@ impl JournalDB {
 			if backing.get(&h.bytes()).expect("Low-level database error. Some issue with your hard disk?").is_some() {
 				// already in the backing DB. start counting, and remember it was already in.
 				Self::set_already_in(batch, &h);
-				counters.insert(h.clone(), 1);
+				refs.insert(h.clone(), RefInfo{queue_refs: 1, in_archive: true});
+				if trace {
```
Comment (arkpar) [inline util/src/journaldb.rs]: `trace!` macro has this check internally. There is no performance penalty when tracing is disabled

**Pull request merged**